clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -182,6 +215,234 @@ public CompletableFuture<Map<TopicPartition, OffsetAndTimestampInternal>> fetchO
+                result.complete(null);
+            } else {
+                result.completeExceptionally(error);
+            }
+        });
+    }
+
+    /**
+     * Fetch the committed offsets for partitions that require initialization. Use them to set
+     * the fetch positions in the subscription state.
+     *
+     * @throws TimeoutException If offsets could not be retrieved within the timeout
+     */
+    private CompletableFuture<Void> initWithCommittedOffsetsIfNeeded(long deadlineMs) {
+        final Set<TopicPartition> initializingPartitions = subscriptionState.initializingPartitions();
+
+        if (initializingPartitions.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        log.debug("Refreshing committed offsets for partitions {}", initializingPartitions);
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        // The shorter the timeout provided to poll(), the more likely the offsets fetch will time out. To handle
+        // this case, on the first attempt to fetch the committed offsets, a FetchCommittedOffsetsEvent is created
+        // (with potentially a longer timeout) and stored. The event is used for the first attempt, but in the
+        // case it times out, subsequent attempts will also use the event in order to wait for the results.
+        CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> refreshWithCommitted;
+        if (!canReusePendingOffsetFetchEvent(initializingPartitions)) {
+            // Generate a new OffsetFetch request and update positions when a response is received
+            final long fetchCommittedDeadlineMs = Math.max(deadlineMs, time.milliseconds() + defaultApiTimeoutMs);
+            CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> fetchOffsets =
+                    commitRequestManager.fetchOffsets(initializingPartitions, fetchCommittedDeadlineMs);
+            refreshWithCommitted = refreshOffsetsAndCompleteResultOnResponseReceived(fetchOffsets, result);
+            pendingOffsetFetchEvent = new PendingFetchCommittedRequest(initializingPartitions, refreshWithCommitted);
+        } else {
+            // Reuse pending OffsetFetch request
+            refreshWithCommitted = pendingOffsetFetchEvent.result;
+            refreshWithCommitted.whenComplete((__, error) -> {
+                if (error == null) {
+                    result.complete(null);
+                } else {
+                    result.completeExceptionally(error);
+                }
+            });
+        }
+
+        return result;
+    }
+
+    private CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> refreshOffsetsAndCompleteResultOnResponseReceived(
+            final CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> fetchCommittedFuture,
+            final CompletableFuture<Void> result) {
+        return fetchCommittedFuture.whenComplete((offsets, error) -> {
+            pendingOffsetFetchEvent = null;
```

Comment:
if `fetchCommittedFuture` is a already completed future, `whenComplete` will be evaluated before return. That means `pendingOffsetFetchEvent` will never be null.

Reply:
Yes, this changed shape with the latest changes, so I believe we don't have such situation anymore. Now the pending goes to null [here](https://github.com/apache/kafka/blob/2b5de7dbc5a5ac9a8dad1bf99254dce8b09e11d0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L372) (even if the OffsetFetch future is somehow done already).